tailnet/configmaps.go

```diff
@@ -387,33 +408,78 @@ func (c *configMaps) updatePeerLocked(update *proto.CoordinateResponse_PeerUpdat
 		// to avoid random hangs while we set up the connection again after
 		// inactivity.
-		node.KeepAlive = ok && peerStatus.Active
+		node.KeepAlive = (statusOk && peerStatus.Active) || (peerOk && lc.node != nil && lc.node.KeepAlive)
 	}
 	switch {
-	case !ok && update.Kind == proto.CoordinateResponse_PeerUpdate_NODE:
+	case !peerOk && update.Kind == proto.CoordinateResponse_PeerUpdate_NODE:
 		// new!
 		var lastHandshake time.Time
 		if ps, ok := status.Peer[node.Key]; ok {
 			lastHandshake = ps.LastHandshake
 		}
-		c.peers[id] = &peerLifecycle{
+		lc = &peerLifecycle{
 			peerID:        id,
 			node:          node,
 			lastHandshake: lastHandshake,
 			lost:          false,
+			// If we're receiving a NODE update for a peer we don't already have
+			// a lifecycle for, it's likely the source of a tunnel. We don't
+			// need to wait for a READY_FOR_HANDSHAKE.
+			readyForHandshake: true,
 		}
+		c.peers[id] = lc
 		logger.Debug(context.Background(), "adding new peer")
-		return true
-	case ok && update.Kind == proto.CoordinateResponse_PeerUpdate_NODE:
+		// since we just got this node, we don't know if it's ready for
+		// handshakes yet.
+		return lc.readyForHandshake
+	case peerOk && update.Kind == proto.CoordinateResponse_PeerUpdate_NODE:
 		// update
-		node.Created = lc.node.Created
-		dirty = !lc.node.Equal(node)
+		if lc.node != nil {
+			node.Created = lc.node.Created
+		}
+		dirty = !lc.node.Equal(node) && lc.readyForHandshake
 		lc.node = node
 		lc.lost = false
 		lc.resetTimer()
+		if lc.isDestination {
+			if !lc.readyForHandshake {
+				lc.setReadyForHandshakeTimer(c)
+			} else {
+				lc.node.KeepAlive = true
+			}
+		}
 		logger.Debug(context.Background(), "node update to existing peer", slog.F("dirty", dirty))
 		return dirty
-	case !ok:
+	case peerOk && update.Kind == proto.CoordinateResponse_PeerUpdate_READY_FOR_HANDSHAKE:
+		dirty := !lc.readyForHandshake
+		lc.readyForHandshake = true
+		if lc.readyForHandshakeTimer != nil {
+			lc.readyForHandshakeTimer.Stop()
+			lc.readyForHandshakeTimer = nil
+		}
+		if lc.node != nil {
+			dirty = dirty || !lc.node.KeepAlive
+			lc.node.KeepAlive = true
+		}
+		logger.Debug(context.Background(), "peer ready for handshake")
+		// only force a reconfig if the node populated
+		return dirty && lc.node != nil
+	case !peerOk && update.Kind == proto.CoordinateResponse_PeerUpdate_READY_FOR_HANDSHAKE:
+		// When we receive a READY_FOR_HANDSHAKE for a peer we don't know about,
+		// we create a peerLifecycle with the peerID and set readyForHandshake
+		// to true. Eventually we should receive a NODE update for this peer,
+		// and it'll be programmed into wireguard.
+		logger.Debug(context.Background(), "got peer ready for handshake for unknown peer")
+		lc = &peerLifecycle{
+			peerID:            id,
+			lost:              true,
+			readyForHandshake: true,
+		}
+		// Timeout the peer in case we never receive a NODE update for it.
+		lc.setLostTimer(c)
```

Comment:
I don't think we should set the Lost timer here, because the peer is not lost.  If we maintain connection to the coordinator, it will tell us when the peer is lost.  If we lose connection to the coordinator, we set all peers to lost.